src/main/java/io/vertx/core/http/impl/Http2UpgradedClientConnection.java

```diff
@@ -0,0 +1,389 @@
+/*
+ * Copyright (c) 2011-2018 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.core.http.impl;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.http.*;
+import io.vertx.codegen.annotations.Nullable;
+import io.vertx.core.*;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.*;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.HttpVersion;
+import io.vertx.core.http.impl.pool.ConnectionListener;
+import io.vertx.core.impl.ContextInternal;
+import io.vertx.core.net.NetSocket;
+import io.vertx.core.net.SocketAddress;
+
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.SSLSession;
+import javax.security.cert.X509Certificate;
+
+/**
+ * An HTTP/2 connection in clear text that upgraded from an HTTP/1 upgrade.
+ */
+public class Http2UpgradedClientConnection implements HttpClientConnection {
```

Comment:
we need to wrap because the `HttpConnection` returned by `connection()` needs to be `this` since the real connection is the `current` field that is initially an `Http1xClientConnection` and then will be an `Http2ClientConnection`.

when the user has set a `HttpClientRequest#connectionHandler` he will receive this instance and keep a reference to it, if we pass directly the HTTP/1 connection then the reference will be invalid after upgrade.